### ISSUE TITLE
RSA-PSS does not have a default digest in master

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#13957]
In 1.1.1, performing a digest-sign operation without specifying a digest will cause OpenSSL to attempt to use the default digest for the specified algorithm. This works fine for RSA-PSS in that branch.

However, in master the logic does not seem to be there - and that seems deliberate for some reason.

The logic for 1.1.1 for determining the default digest is here:

https://github.com/openssl/openssl/blob/a7222fc14d5977210d2b4673a68039824a039dc2/crypto/rsa/rsa_ameth.c#L511-L523

In master this is where we might expect it to be handled:

https://github.com/openssl/openssl/blob/b897b353dff8138aa838bae9766ecd3de8c03280/providers/implementations/keymgmt/rsa_kmgmt.c#L314-L322

As you can see there is deliberate logic and a comment that provides a default digest for normal RSA, but for RSA-PSS it is skipped.

This is the cause of the failure noted here in https://github.com/openssl/openssl/issues/13931#issuecomment-766897892

**Comment by t8m** [comment]
I do not see a reason why RSA-PSS key with ossl_rsa_pss_params_30_is_unrestricted should not return a default digest same as regular RSA key.

**Comment by levitte** [comment]
What you're missing in all this is the following:

``` C
         /* Return of 2 indicates this MD is mandatory */
         return 2; 
```

Because the set_params / get_params return values are strictly 0 (failure) or 1 (success), we don't have the same possibility to get that kind of indication back.  The solution that was implemented is more visible if you look a little further in the `rsa_keymgmt.c` code:

https://github.com/openssl/openssl/blob/b897b353dff8138aa838bae9766ecd3de8c03280/providers/implementations/keymgmt/rsa_kmgmt.c#L314-L337

In other words, instead of getting a signal back whether the returned digest is mandatory or not, you now have to ask for the mandatory first, and for the default if that fails.

Although....  re https://github.com/openssl/openssl/issues/13931#issuecomment-766897892, I'm still a bit surprised that you got an error, considering that EVP_DigestSignInit does check for the mandatory digest first...  what am I missing?

**Comment by t8m** [comment]
> Although.... re [#13931 (comment)](https://github.com/openssl/openssl/issues/13931#issuecomment-766897892), I'm still a bit surprised that you got an error, considering that EVP_DigestSignInit does check for the mandatory digest first... what am I missing?

Because with RSA_PSS keys for which ossl_rsa_pss_params_30_is_unrestricted() returns true, there is no OSSL_PKEY_PARAM_MANDATORY_DIGEST returned.

**Comment by levitte** [comment]
Ahhhhh!
